File: NeetCodeIO/994RottingOranges/RottingOranges.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        // BFS -> Time:O(mn), Space:O(mn)
        queue<pair<int,int>> q;
        int time = 0, fresh = 0;
        int ROWS = grid.size(), COLS = grid[0].size();
        for (int i = 0; i < ROWS; i++) {
            for (int j = 0; j < COLS; j++) {
                if (grid[i][j] == 2) {
                    pair<int,int> temp(i,j);
                    q.push(temp);
                }
                if (grid[i][j] == 1) {
                    fresh++;
                }
            }
        }
        vector<pair<int,int>> directions = { {1, 0}, {-1, 0}, {0, 1}, {0, -1} };;
        while (!q.empty() && fresh > 0) {
            int n = q.size();
            for (int i = 0; i < n; i++) {
                int row = q.front().first, col = q.front().second;
                q.pop();
                for (int j = 0; j < directions.size(); j++) {
                    int dr = directions[j].first, dc = directions[j].second;
                    int r = row + dr, c = col + dc;
                    if (r < 0 || c < 0 || r >= ROWS || c >= COLS || grid[r][c] != 1) {
                        continue;
                    }
                    grid[r][c] = 2;
                    fresh--;
                    pair<int,int> temp(r,c);
                    q.push(temp);
                }
            }
            time++;
        }
        if (fresh) {
            return -1;
        }
        return time;
    }
};

File: NeetCodeIO/994RottingOranges/RottingOranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        # BFS -> Time:O(mn), Space:O(mn)
        q = []
        time, fresh = 0, 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 2:
                    q.append((i,j))
                if grid[i][j] == 1:
                    fresh += 1
        directions = ((1, 0),(-1, 0),(0, 1),(0, -1))
        while q and fresh > 0:
            for _ in range(len(q)):
                r, c = q.pop(0)
                for dr, dc in directions:
                    row, col = r + dr, c + dc
                    if row < 0 or col < 0 or row >= len(grid) or col >= len(grid[0]) or grid[row][col] != 1:
                        continue
                    q.append((row, col))
                    grid[row][col] = 2
                    fresh -= 1
            time += 1
        return time if fresh == 0 else -1
